LonaStudio/Workspace/ViewController.swift

```diff
@@ -590,20 +604,39 @@ class ViewController: NSViewController, NSOutlineViewDataSource, NSOutlineViewDe
 
         if characters == String(Character(UnicodeScalar(NSDeleteCharacter)!)) {
             guard let targetLayer = selectedLayer else { return }
-
             if targetLayer === dataRoot { return }
-
             let parent = outlineView!.parent(forItem: targetLayer) as! CSLayer
-            parent.children = parent.children.filter({ $0 !== targetLayer })
 
-            clearInspector()
-            renderLayerList()
-            render()
+            // Undo
+            let oldChildren = parent.children
+            let children = parent.children.filter({ $0 !== targetLayer })
+            UndoManager.shared.run(name: "Delete", execute: {[unowned self] in
+                self.updateChildren(children: children, for: parent)
+            }, undo: {[unowned self] in
+                self.updateChildren(children: oldChildren, for: parent)
+            })
         } else if characters == String(Character(" ")) {
             canvasCollectionView?.panningEnabled = true
         }
     }
 
+    private func updateChildren(children: [CSLayer], for parent: CSLayer) {
+
+        // Change
+        parent.children = children
+
+        // Refresh
+        refreshDocument(true)
+    }
+
+    private func refreshDocument(_ shouldClearInspector: Bool = false) {
```

Comment:
In general I'd leave the parameter label on a boolean rather than using `_`, since it's easier to understand when reading a function call. E.g. `refreshDocument(shouldClearInspector: true)` explains what the boolean does